js/depgraph.js

```diff
@@ -0,0 +1,279 @@
+      logger.debug('visiting broken node ', pluginNode)
+      if (pluginNode.visited) {
+        return;
+      } 
+      if (pluginNode.visiting) {
+        //TODO deal with a circular dependency. Not really clear what to do,
+        //perhaps, reject the entire cycle but leave the unaffected nodes there.
+        //Implementing a cycle detection algorithm is a different story.
+        //TODO good diagnostics
+        throw new Error("circular dependency: " + pluginNode.pluginId);
+      }
+      pluginNode.valid = false;
+      pluginNode.validationError = validationError;
+      pluginNode.visiting = true;
+      for (const dep of pluginNode.deps) {
+        const importer = graph[dep.importer];
+        logger.debug("following link: ", dep, ": ", importer)
+        let error;
+        if (!dep.valid) {
+          error = dep.validationError;
+        } else {
+          error = {
+            status: "REQUIRED_PLUGIN_FAILED_TO_LOAD",
+            pluginId: dep.provider
+          }
+        }
+        visit(importer, error);
+      }
+      rejects[pluginNode.pluginId] = pluginNode;
+      pluginNode.visiting = false;
+      pluginNode.visited = true;
+    }
+    for (const reject of Object.keys(rejects)) {
+      delete graphWithBrokenDeps.graph[reject.pluginId];
+    }
+    return rejects;
+  },
+  
+  /**
+   * Produces a topologically sorted array of plugins. 
+   * 
+   * Note that the sorted list returned can contain "dream" plugins - those that 
+   * were required but don't exist. The caller needs to filter them out (the
+   * importers will of course be correctly rejected)
+   * 
+   */
+  _toposort(graph) {
+    logger.debug('graph: ', graph)
+    const pluginsSorted = [];
+    let time = 0;
+    for (let importedPlugin of Object.values(graph)) {
+      visit(importedPlugin, true);
+    }
+    return pluginsSorted;
+    
+    function visit(pluginNode) {
+      logger.debug('visiting node ', pluginNode)
+      if (pluginNode.visited) {
+        return;
+      } 
+      if (pluginNode.visiting) {
+        //TODO deal with a circular dependency. Not really clear what to do,
+        //perhaps, reject the entire cycle but leave the unaffected nodes there.
+        //Implementing a cycle detection algorithm is a different story.
+        //TODO good diagnostics
+        throw new Error("circular dependency: " + pluginNode.pluginId);
+      }
+      pluginNode.discoveryTime = ++time;
+      pluginNode.visiting = true;
+      for (const dep of pluginNode.deps) {
+        visit(graph[dep.importer]);
+      }
+      pluginNode.visiting = false;
+      pluginNode.visited = true;
+      pluginNode.finishingTime = ++time;
+      logger.debug(`${pluginNode.pluginId}: `
+          + `${pluginNode.discoveryTime}/${pluginNode.finishingTime}`);
+      //See the proof at the end of Cormen et al. (2001), 
+      // "Section 22.4: Topological sort"
+      //loop invariant derived from the proof
+      assert((pluginsSorted.length === 0) 
```

Comment:
Perhaps it will be better to comment it out in production. the thing with assert in JS is that an AssertionError is just an error, that will be caught by *any* catch block. So, most likely, it will just confuse the program in some place and won't terminate the process anyway.